File: src/python/transaction/working_with_transaction/reading_transaction_properties/reading_properties.py
import asyncio

from wax import create_hive_chain
from wax.proto.operations import vote


# Initialize hive chain interface
chain = create_hive_chain()


async def main() -> None:
    # Initialize a transaction object
    tx = await chain.create_transaction()

    # Declare example operation

    vote_operation = vote(voter="voter", author="test-author", permlink="test-permlink", weight=2200)

    # Push operation into the transaction
    tx.push_operation(vote_operation)

    # Get a transaction object holding all operations and transaction TAPOS & expiration data, but transaction is **not signed yet**
    built_transaction = tx.transaction

    # Most transaction properties should be read from the transaction before building it.
    print(f"id: {tx.id}")
    print(f"sig_digest: {tx.sig_digest}")

    # Some transaction properties should be read from the transaction after building it.
    print(f"expiration: {built_transaction.expiration}")
    print(f"ref_block_num: {built_transaction.ref_block_num}")


asyncio.run(main())

File: src/rust/transaction/working_with_transaction/reading_transaction_properties/reading_properties.rs
use wax::prelude::*;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Initialize hive chain interface
    let chain = create_hive_chain(None)?;

    // Initialize a transaction object
    let mut tx = chain.create_transaction(None).await?;

    // Declare example operation
    let operation = chain.create_operation(Value::VoteOperation(Vote {
        voter: "voter".into(),
        author: "test-author".into(),
        permlink: "test-permlink".into(),
        weight: 2200,
    }));

    // Push operation into the transaction
    tx.push_operation(operation);

    /*
    Get a transaction object holding all operations and transaction
    TAPOS & expiration data, but transaction is **not signed yet**
    */
    let built_transaction = tx.transaction();

    println!("id: {}", tx.id()?);
    println!("sigDigest: {}", tx.sig_digest()?);
    println!("expiration: {}", built_transaction.expiration);
    println!("ref_block_num: {}", built_transaction.ref_block_num);

    Ok(())
}
